fix(linter): false positive with negative matches in no-restricted-imports

camc314 opened the pull request:
this was an absolute nightmare to fix.
TLDR is that `GitignoreBuilder` nor `OverrideBuilder` do not contain the flexibility or have the behaviour we want to achieve this here. They don't go through the full list of patterns (for perf reasons), but that has the caveat of meaning it does not work correctly with eslint

closes #10911

camc314 commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10976?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#10976** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10976?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10976?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#10975** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10975?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=10976).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Sysix commented on crates/oxc_linter/src/rules/eslint/no_restricted_imports.rs:
```diff
@@ -856,30 +857,37 @@ impl RestrictedPattern {
+            };
+
+            if glob.is_match(name) {
+                decision = if negated { GlobResult::Whitelist } else { GlobResult::Found };
```
What if we iterate over multiple groups? The second iteration will override maybe Whitelist and report an unwanted error?
I think we should fast return for `GlobResult::Whitelist`

camc314 replied:
let me work out what eslint does

camc314 replied:
added another test, our behaviour should be alligned with eslint now:

https://eslint.org/play/#eyJ0ZXh0IjoiLy8gKFxuLy8gICAgICAgICAgICAgciNcImltcG9ydCBhIGZyb20gXCIuL2luZGV4Lm1qc1wiO1wiIyxcbi8vICAgICAgICAgICAgIFNvbWUoXG4vLyAgICAgICAgICAgICAgICAgc2VyZGVfanNvbjo6anNvbiEoW3sgXCJwYXR0ZXJuc1wiOiBbeyBcImdyb3VwXCI6IFtcIltAYS16XSpcIiwgXCIhLiovKipcIiwgXCIuL2luZGV4Lm1qc1wiXSwgXCJtZXNzYWdlXCI6IFwiZm9vIGlzIGZvcmJpZGRlbiwgdXNlIGJhciBpbnN0ZWFkXCIgfV0gfV0pLFxuLy8gICAgICAgICAgICAgKSxcbi8vICAgICAgICAgKSxcblxuaW1wb3J0IFwiZm9vXCI7XG5pbXBvcnQgXCIuL2luZGV4Lm1qc1wiO1xuaW1wb3J0IFwiaW5kZXhcIjsiLCJvcHRpb25zIjp7InJ1bGVzIjp7Im5vLXJlc3RyaWN0ZWQtaW1wb3J0cyI6WyJlcnJvciIseyJwYXR0ZXJucyI6W3siZ3JvdXAiOlsiW0BhLXpdKiIsIiEuKi8qKiIsIi4vaW5kZXgubWpzIl0sIm1lc3NhZ2UiOiJmb28gaXMgZm9yYmlkZGVuLCB1c2UgYmFyIGluc3RlYWQifV19XX0sImxhbmd1YWdlT3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e319LCJlY21hVmVyc2lvbiI6ImxhdGVzdCJ9fX0=

Sysix commented:
### Merge activity

* **May 12, 3:03 PM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

camc314 committed:
fix(linter): false positive with negative matches in no-restricted-imports (#10976)

this was an absolute nightmare to fix.
TLDR is that `GitignoreBuilder` nor `OverrideBuilder` do not contain the flexibility or have the behaviour we want to achieve this here. They don't go through the full list of patterns (for perf reasons), but that has the caveat of meaning it does not work correctly with eslint

closes #10911
